api/envoy/extensions/filters/http/stateful_session/v3/stateful_session.proto

```diff
@@ -0,0 +1,55 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.stateful_session.v3;
+
+import "envoy/config/core/v3/extension.proto";
+import "envoy/config/core/v3/health_check.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.stateful_session.v3";
+option java_outer_classname = "StatefulSessionProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/stateful_session/v3;stateful_sessionv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Stateful session filter]
+// Stateful session :ref:`configuration overview <config_http_filters_stateful_session>`.
+// [#extension: envoy.filters.http.stateful_session]
+
+// [#comment:TODO(wbpcode): To support global feature enabled or disabled flag.]
+message StatefulSession {
+  // The expected health status of the host. Suppose a session is assigned to the host X.
+  // However, if the health status of host X does not meet expectations, Envoy will select
+  // new host for the session according to the load balancing algorithm. In other words,
+  // the status of the target host must be in this list for session sticky to take effect.
+  // If no expected health status is configured then any host status will be accepted by
+  // default.
+  //
+  // .. note::
+  //
+  //    The *DRAINING* and *TIMEOUT* will be interpreted as *UNHEALTHY* and *UNKNOWN* will
+  //    be interpreted as *HEALTHY*.
+  repeated config.core.v3.HealthStatus host_statuses = 1;
```

Comment:
This feels to be more like `Cluster`-level configuration. We're basically talking about how host picking works. I think `session_state` is fine in HTTP filter, because this is how we are encoding and managing information we return to the client and receive from it. But the actual LB piece should be guided from LB policy.

Reply:
🤔 This is an optional configuration. With this configuration, we can stop the'hard' sticky and switch to relying on lb to select a new host when the status of the host does not match. 

This actually corresponds to the API of `LoadBalancerContext`. We use this API to influence the behaviour of LBs and ultimately achieve hard sticky.

```
  // Using uint32_t to express expected status of override host. Every bit in the OverrideHostStatus
  // represent an enum value of Host::Health. The specific correspondence is shown below:
  //
  // * 0b001: Host::Health::Unhealthy
  // * 0b010: Host::Health::Degraded
  // * 0b100: Host::Health::Healthy
  //
  // If multiple bit fields are set, it is acceptable as long as the status of override host is in
  // any of these statuses.
  using OverrideHostStatus = uint32_t;
  using OverrideHost = std::pair<std::string, OverrideHostStatus>;

  /**
   * Returns the host the load balancer should select directly. If the expected host exists and
   * the health status of the host matches the expectation, the load balancer can bypass the load
   * balancing algorithm and return the corresponding host directly.
   */
  virtual absl::optional<OverrideHost> overrideHostToSelect() const PURE;
```

Reply:
Yeah, I understand how it's been plumbed here, but it doesn't change the fact that there really is no need to plumb stickiness settings from HTTP filters across to CM. HTTP filters should be concerned with extracting HTTP state and setting it in a pluggable way for hard sticky. The rest of the configuration, including what constitutes a valid LB host based on health, is really an LB setting that should logically belong to `Cluster` IMHO.

I might be expressing just my take here, so I would be interested in others' opinions. Either @alyssawilk who has reviewed your doc or @markdroth who is also going to be interested in this API proposal as it will influence hard stickiness for gRPC LB.

Reply:
@htuch I probably understand what you mean. In this case, we just need to tell LB what we want the host to be. The LB decides if the current state of the target host is acceptable for 'hard' sticky. right?

Reply:
Yep, exactly. It's something I think we should at least consider as an alternative here to the way the API is structured. I don't think it changes a lot of code for you, in fact it should make things simpler (no need to plumb across the health criteria from HTTP filter).

Reply:
sound reasonable. 🤔 It's just a little change for this PR. But we need a separate PR  to update `LoadBalancerContext` interface and also the cluster config.